alfresco-zipkin-repo/src/main/java/eu/xenit/alfresco/instrumentation/solr/VirtualSolrSpanFactory.java

```diff
@@ -0,0 +1,221 @@
+        mainTimings = getTimingsFromResponseJson(jsonResponse);
+        solrRequest = new SolrRequest(numFound, qTime, query, mainTimings);
+    }
+
+    public Span createVirtualSolrSpans(Tracer tracer, Span solrSpan, long startTime, long endTime) {
+
+        //Apply main solr information to span
+        solrRequest.applyToSpan(solrSpan, startTime);
+
+        //Add sharded boolean tag
+        solrSpan.tag("sharded", String.valueOf(sharded));
+
+        //Approximation of time that main solr request was not actively searching -> I/O and package transmissions
+        long transmissionTime = ((endTime - startTime) - solrRequest.qTime * 1000) / 2;
+
+        if (shardedRequests != null) {
+            //Find the first solr shard instance to which the request was sent to
+            //Example path /solr/shard1/afts
+            String shardInstanceIdentifier = getShardInstanceIdentifier(receivingSolrInstance, 2);
+            ShardedSolrRequest firstRequest = shardedRequests.get(shardInstanceIdentifier);
+
+            //Create a new span for the first sharded request
+            Span firstShardedSpan = tracer.newChild(solrSpan.context());
+
+            //Start the first span on startTime of main span + transmission time to first shard instance (approximation)
+            long firstReceiveTime = startTime + transmissionTime;
+
+            //Apply to newly create span for the first shard instance which propagates the query to the other shards
+            firstRequest.completeSpan(firstShardedSpan, firstReceiveTime);
+
+            //Create a child span for the different subrequests
+            firstRequest.createAndCompleteSubSpans(tracer, firstShardedSpan, firstReceiveTime);
+
+            //Approximate transmission time between different shard instances (similar approx)
+            long shardInstanceTransmissionTime = ((firstRequest.elapsedTime - firstRequest.qTime) / 3) * 1000;
+
+            for (ShardedSolrRequest shardedSolrRequest : shardedRequests.values()) {
+                if (shardedSolrRequest != firstRequest) {
+                    //Create a new span for each further shardedRequest
+                    Span subsequentSpan = tracer.newChild(solrSpan.context());
+                    long spanStartTime = firstReceiveTime + shardInstanceTransmissionTime;
+                    shardedSolrRequest.completeSpan(subsequentSpan, spanStartTime);
+                    shardedSolrRequest.createAndCompleteSubSpans(tracer, subsequentSpan, spanStartTime);
+                }
+            }
+        }
+
+        return solrSpan;
+    }
+
+    private String getShardInstanceIdentifier(String in, int index) {
+        String[] segments = in.split("/");
+        if (index == -1) {
+            return segments[segments.length - 1];
+        } else {
+            return segments[index];
+        }
+    }
+
+    private JSONObject getTimingsFromResponseJson(JSONObject jsonResponse) {
+        if (jsonResponse != null && sharded == false) {
```

Comment:
You're calling this function from the constructor, and you're relying on some fields already being set before your function is called.

It's a bit dangerous to rely on that, as people may move code in the constructor around, which would suddenly make this function fail.